### ISSUE TITLE
Duplicate SDPA Input Constraints

### ISSUE THREAD

**Issue by kqfu** [pytorch/pytorch#175833]
When applying constraints to SDPA, if K==V, it's doing duplicate copy.

Small repro where you see 2 identical empty_strided_cuda:
```
import torch
import torch._inductor.config as inductor_config
from torch._inductor.utils import run_and_get_code


def fn(q, kv):
    kv_act = torch.relu(kv)
    return torch.nn.functional.scaled_dot_product_attention(q, kv_act, kv_act)


q = torch.randn(2, 2, 16, 32, device="cuda", dtype=torch.float16)
kv = torch.randn(2, 2, 16, 32, device="cuda", dtype=torch.float16)

_, (code,) = run_and_get_code(torch.compile(fn), q, kv)
print(code)
```

cc @jerryzh168 @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by karthickai** [comment]
**🤖 PT2 Repro Agent**

I tried to reproduce this issue on PyTorch nightly (`2.13.0.dev20260516+cu132`) on NVIDIA H100 but could not reproduce it (all 3 runs passed).

All 3 runs of the issue's exact repro found zero duplicate `empty_strided_cuda` calls in the generated inductor code. The generated wrapper allocates a single `buf0` for `relu(kv)` and passes the same buffer to both the K and V slots of `_scaled_dot_product_cudnn_attention(arg1_1, buf0, buf0, None, False)`, so no duplicate copy is emitted.

<details>
<summary>Error output</summary>

```
PyTorch: 2.13.0.dev20260516+cu132
CUDA:    True
GPU:     NVIDIA H100
=== Generated code (first 4000 chars) ===
# AOT ID: ['0_inference']
from ctypes import c_void_p, c_long, c_int
import torch
import math
import random
import os
import tempfile
from math import inf, nan
from cmath import nanj
from torch._inductor.hooks import run_intermediate_hooks
from torch._inductor.utils import maybe_profile
from torch._inductor.codegen.memory_planning import _align as align
from torch import device, empty_strided
from torch._inductor.async_compile import AsyncCompile
from torch._inductor.select_algorithm import extern_kernels
from torch._C._dynamo.guards import copy_if_misaligned
import triton
import triton.language as tl
from torch._inductor.runtime.triton_heuristics import start_graph, end_graph
from torch._C import _cuda_getCurrentRawStream as get_raw_stream

aten = torch.ops.aten
inductor_ops = torch.ops.inductor
_quantized = torch.ops._quantized
assert_size_stride = torch._C._dynamo.guards.assert_size_stride
assert_alignment = torch._C._dynamo.guards.assert_alignment
empty_strided_cpu = torch._C._dynamo.guards._empty_strided_cpu
empty_strided_cpu_pinned = torch._C._dynamo.guards._empty_strided_cpu_pinned
empty_strided_cuda = torch._C._dynamo.guards._empty_strided_cuda
empty_strided_xpu = torch._C._dynamo.guards._empty_strided_xpu
empty_strided_mtia = torch._C._dynamo.guards._empty_strided_mtia
reinterpret_tensor = torch._C._dynamo.guards._reinterpret_tensor
alloc_from_pool = torch.ops.inductor._alloc_from_pool
async_compile = AsyncCompile()
empty_strided_p2p = torch._C._distributed_c10d._SymmetricMemory.empty_strided_p2p


# kernel path: /tmp/torchinductor_karthickps/vr/cvryrsswsvzq4w332f5ftspazqc5ikmx6y5qtffsswugapeeknl7.py
# Topologically Sorted Source Nodes: [kv_act], Original ATen: [aten.relu]
# Source node to ATen node mapping:
#   kv_act => relu
# Graph fragment:
#   %arg0_1 : Tensor "f16[2, 2, 16, 32][1024, 512, 32, 1]cuda:0" = PlaceHolder[target=arg0_1]
#   %relu : Tensor "f16[2, 2, 16, 32][1024, 512, 32, 1]cuda:0"[num_users=1] = call_function[target=torch.ops.aten.relu.default](args = (%arg0_1,), kwargs = {})
#   return %relu
triton_poi_fused_relu_0 = async_compile.triton('triton_poi_fused_relu_0', '''
import triton
import triton.language as tl

from torch._inductor.runtime import triton_helpers, triton_heuristics
from torch._inductor.runtime.triton_helpers import libdevice, math as tl_math
from torch._inductor.runtime.hints import AutotuneHint, ReductionHint, TileHint, DeviceProperties
triton_helpers.set_driver_to_gpu()

@triton_heuristics.pointwise(
    size_hints={'x': 2048}, 
    filename=__file__,
    triton_meta={'signature': {'in_ptr0': '*fp16', 'out_ptr0': '*fp16', 'xnumel': 'i32', 'XBLOCK': 'constexpr'}, 'device': DeviceProperties(type='cuda', index=0, multi_processor_count=132, cc=90, major=9, regs_per_multiprocessor=65536, max_threads_per_multi_processor=2048, max_threads_per_block=1024, warp_size=32), 'constants': {}, 'native_matmul': False, 'enable_fp_fusion': True, 'launch_pdl': False, 'disable_ftz': False, 'configs': [{(0,): [['tt.divisibility', 16]], (1,): [['tt.divisibility', 16]], (2,): [['tt.divisibility', 16]]}]},
    inductor_meta={'grid_type': 'Grid1D', 'kernel_name': 'triton_poi_fused_relu_0', 'mutated_arg_names': [], 'optimize_mem': True, 'no_x_dim': False, 'atomic_add_found': False, 'num_load': 1, 'num_store': 1, 'num_reduction': 0, 'autotune_hints': set(), 'tiling_scores': {'x': 12288}, 'backend_hash': '80C12874F41CD40C1C5FC8D981BC028920B3B1AFED1F7C49746F45561A21E10F', 'assert_indirect_indexing': True, 'autotune_local_cache': True, 'autotune_pointwise': True, 'autotune_remote_cache': None, 'force_disable_caches': False, 'dynamic_scale_rblock': True, 'incremental_autotune': False, 'max_autotune': False, 'max_autotune_pointwise': False, 'min_split_scan_rblock': 256, 'spill_threshold': 16, 'store_cubin': False, 'deterministic': False, 'batch_invariant': False, 'force_filter_reduction_configs': False, 'mix_order_reduction_allow_multi_stages': True, 'dynamic_disable_pipelining': True, 'are_deterministic_algorithm
=== Duplicate empty_strided_cuda calls ===
  (none)


```
</details>

<details>
<summary>Repro script I used</summary>

```python
import os
import re
from collections import Counter

import torch
from torch._inductor.utils import run_and_get_code


def fn(q, kv):
    kv_act = torch.relu(kv)
    return torch.nn.functional.scaled_dot_product_attention(q, kv_act, kv_act)


def _find_duplicate_allocations(code: str):
    """Find empty_strided_cuda(...) calls that appear more than once."""
    pattern = re.compile(r"empty_strided_cuda\([^)]*\)")
    allocations = pattern.findall(code)
    counts = Counter(allocations)
    return [(s, n) for s, n in counts.items() if n > 1]


def test_fn() -> bool:
    """Return True if NO bug. False if duplicate empty_strided_cuda calls exist."""
    torch._dynamo.reset()
    backend = os.environ.get("TORCH_COMPILE_BACKEND", "inductor")

    q = torch.randn(2, 2, 16, 32, device="cuda", dtype=torch.float16)
    kv = torch.randn(2, 2, 16, 32, device="cuda", dtype=torch.float16)

    compiled = torch.compile(fn, backend=backend)

    # For non-inductor backends, there's no inductor wrapper code to inspect.
    # The bug is in inductor codegen specifically — exercise the path and pass.
    if backend != "inductor":
        compiled(q, kv)
        return True

    _, codes = run_and_get_code(compiled, q, kv)
    code = "\n".join(codes) if codes else ""
    duplicates = _find_duplicate_allocations(code)
    return len(duplicates) == 0


if __name__ == "__main__":
    print(f"PyTorch: {torch.__version__}")
    print(f"CUDA:    {torch.cuda.is_available()}")
    if torch.cuda.is_available():
        print(f"GPU:     {torch.cuda.get_device_name(0)}")

    # Diagnostic: re-run to dump the generated code & the duplicates we found
    torch._dynamo.reset()
    q = torch.randn(2, 2, 16, 32, device="cuda", dtype=torch.float16)
    kv = torch.randn(2, 2, 16, 32, device="cuda", dtype=torch.float16)
    compiled = torch.compile(fn)
    _, codes = run_and_get_code(compiled, q, kv)
    code = "\n".join(codes) if codes else ""
    print("=== Generated code (first 4000 chars) ===")
    print(code[:4000])
    print("=== Duplicate empty_strided_cuda calls ===")
    dups = _find_duplicate_allocations(code)
    if dups:
        for s, n in dups:
            print(f"  count={n}: {s}")
    else:
        print("  (none)")

    try:
        ok = test_fn()
    except Exception:
        import traceback
        traceback.print_exc()
        os._exit(2)
    os._exit(0 if ok else 1)

```
</details>

<details>
<summary>My environment</summary>

```
Collecting environment information...
PyTorch version: 2.13.0.dev20260516+cu132
Is debug build: False
CUDA used to build PyTorch: 13.2
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-14)
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.34

Python version: 3.13.12 | packaged by Anaconda, Inc. | (main, Feb 24 2026, 16:13:31) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.16.1-0_fbk2_hardened_0_gf40efc324cc8-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA H100
Nvidia driver version: 580.126.09
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  46
On-line CPU(s) list:                     0-45
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 9654 96-Core Processor
CPU family:                              25
Model:                                   17
Thread(s) per core:                      1
Core(s) per socket:                      46
Socket(s):                               1
Stepping:                                1
BogoMIPS:                                4792.79
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm rep_good nopl cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw perfctr_core ssbd ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512_bf16 clzero xsaveerptr wbnoinvd arat npt lbrv nrip_save tsc_scale vmcb_clean flushbyasid pausefilter pfthreshold v_vmsave_vmload vgif vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid fsrm flush_l1d arch_capabilities
Virtualization:                          AMD-V
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               2.9 MiB (46 instances)
L1i cache:                               2.9 MiB (46 instances)
L2 cache:                                23 MiB (46 instances)
L3 cache:                                736 MiB (46 instances)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-45
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Vulnerable
Vulnerability Spec store bypass:         Vulnerable
Vulnerability Spectre v1:                Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:                Vulnerable; IBPB: disabled; STIBP: disabled; PBRSB-eIBRS: Not affected; BHI: Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.4.4
[pip3] nvidia-cublas==13.4.0.1
[pip3] nvidia-cuda-cupti==13.2.75
[pip3] nvidia-cuda-nvrtc==13.2.78
[pip3] nvidia-cuda-runtime==13.2.75
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.2.0.46
[pip3] nvidia-curand==10.4.2.55
[pip3] nvidia-cusolver==12.2.0.1
[pip3] nvidia-cusparse==12.7.10.1
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.2.78
[pip3] nvidia-nvtx==13.2.75
[pip3] torch==2.13.0.dev20260516+cu132
[pip3] torchao==0.17.0
[pip3] torchaudio==2.11.0.dev20260516+cu132
[pip3] torchdata==0.11.0
[pip3] torchmetrics==1.0.3
[pip3] torchrec==1.7.0.dev20260413+cu130
[pip3] torchtune==0.6.1
[pip3] torchvision==0.28.0.dev20260516+cu132
[pip3] triton==3.7.0+git88b227e2
[conda] Could not collect
```
</details>

---
If the repro agent produced an incorrect result (wrong classification, bad repro script, missed the actual bug, etc.), add the label `repro:agent:h100:incorrect` to the issue.

**Comment by yushangdi** [comment]
I can still repro this on `2.14.0a0+git7e63a83`, but it can be avoided with `TORCHINDUCTOR_CACHE_SDPA_CONSTRAINT=1`.

**Comment by drisspg** [comment]
@yushangdi tbh dont know what this is but is `TORCHINDUCTOR_CACHE_SDPA_CONSTRAINT` the intended fix? if so can we close this issue?

**Comment by yushangdi** [comment]
@drisspg  from https://github.com/pytorch/pytorch/pull/175599, looks like it's a temporary fix. The long term fix should be some better CSE pass in inductor. 

close, feel free to re-open.
